pkg/render/intrusiondetection/dpi/dpi.go

```diff
@@ -0,0 +1,391 @@
+		ids:              ids,
+		installation:     installation,
+		nodeTLSSecret:    nodeTLSSecret,
+		typhaTLSSecret:   typhaTLSSecret,
+		typhaCAConfigMap: typhaCAConfigMap,
+		pullSecrets:      pullSecrets,
+		openshift:        openshift,
+		hasNoLicense:     hasNoLicense,
+	}
+}
+
+type dpi struct {
+	ids              *operatorv1.IntrusionDetection
+	installation     *operatorv1.InstallationSpec
+	nodeTLSSecret    *corev1.Secret
+	typhaTLSSecret   *corev1.Secret
+	typhaCAConfigMap *corev1.ConfigMap
+	pullSecrets      []*corev1.Secret
+	openshift        bool
+	hasNoLicense     bool
+	dpiImage         string
+}
+
+func (d *dpi) ResolveImages(is *operatorv1.ImageSet) error {
+	reg := d.installation.Registry
+	path := d.installation.ImagePath
+	prefix := d.installation.ImagePrefix
+	var err error
+	d.dpiImage, err = components.GetReference(components.ComponentDeepPacketInspection, reg, path, prefix, is)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *dpi) Objects() (objsToCreate, objsToDelete []client.Object) {
+	var toDelete []client.Object
+	var toCreate []client.Object
+	// Add DeepPacketInspection related objects if all necessary bit of TLS config are set.
+	addDPIDaemonSet := d.nodeTLSSecret != nil && d.typhaTLSSecret != nil && d.typhaCAConfigMap != nil
+
+	// toCreate is populated only if DPI daemonset needs to be created and there is valid license,
+	// for all other combinations delete the DPI namespace.
+	if addDPIDaemonSet && !d.hasNoLicense {
+		// Update the IDS resource with the default values for DPI ComponentResources if it doesn't exist,
+		// these values are used for setting container resource requirements in DPI daemonset.
+		d.defaultIntrusionDetectionComponentResources()
+		toCreate = append(toCreate, d.ids)
+
+		toCreate = append(toCreate, render.CreateNamespace(DeepPacketInspectionNamespace, d.installation.KubernetesProvider))
+		toCreate = append(toCreate, secret.ToRuntimeObjects(secret.CopyToNamespace(DeepPacketInspectionNamespace, d.nodeTLSSecret)...)...)
+		toCreate = append(toCreate, secret.ToRuntimeObjects(secret.CopyToNamespace(DeepPacketInspectionNamespace, d.typhaTLSSecret)...)...)
+		toCreate = append(toCreate, configmap.ToRuntimeObjects(configmap.CopyToNamespace(DeepPacketInspectionNamespace, d.typhaCAConfigMap)...)...)
+		toCreate = append(toCreate,
+			d.dpiServiceAccount(),
+			d.dpiClusterRole(),
+			d.dpiClusterRoleBinding(),
+			d.dpiDaemonset(),
+		)
+		return toCreate, nil
+	}
+
+	toDelete = append(toDelete, d.dpiNamespace())
+	return nil, toDelete
+}
+
+func (d *dpi) Ready() bool {
+	return true
+}
+
+func (d *dpi) SupportedOSType() meta.OSType {
+	return rmeta.OSTypeLinux
+}
+
+// defaultIntrusionDetectionComponentResources sets the default requirements value for DPI in IDS resource if it is not set.
+func (d *dpi) defaultIntrusionDetectionComponentResources() {
+	if d.ids.Spec.ComponentResources == nil {
+		d.ids.Spec.ComponentResources = []operatorv1.IntrusionDetectionComponentResource{
```

Comment:
We should not be modifying the ids resource in here, if there's defaults to set they should be set in the controller and written back to the k8s API.